ADR/0039-creating-pending-pipelineruns.md

```diff
@@ -0,0 +1,66 @@
+# 39. Creating pending pipeline runs
+
+Date: 2024-11-21
+
+## Status
+
+Accepted
+
+## Context
+
+There is a frequent use case amongst Konflux users when triggering large amount of builds: sometimes users do not care how quickly the pipeline run will finish, but they want it to finish eventually.
+Now what happens is that part of the pipeline runs started in this big bulk is blocked by unavailable resources (e.g. namespace memory limit quotas or MPC pool size).
+Because of how Tekton works, no matter what these pipeline runs are considered running (even if their pods are waiting for resources) and so eventually timeout setting strikes.
+
+So say you started 300 pipeline runs at Fry evening and you expect to see 300 new images at Mon morning.
+You have resource quota for 20 concurrent builds, so 20 builds get started, when they finish 20 more and so on.
+But because all 300 pipeline runs started (from Tekton point of view) when they were created, after few iterations all the remaining pipeline runs will be cancelled (as timeouted).
```

Comment:
As I mentioned above, the case could be that none of the pipelines completed because they were all starving each other of the needed resources.

Reply:
Hello! I have not experienced that, but it is definitely possible, especially in combination with some MPC bugs.

This is mine mental model of what is happening:

Say you have this pipeline:

1. init
2. clone
3. prefetch
4. [build-s390x, build-arm64]

The pipeline usually runs for 1 hour and have timeout set to 2 hours. MPC have capacity of 2 s390x and 2 arm64, all free.

Now you trigger 5 builds.

1. First 2 builds start, all 5 builds are considered started by Tekton, last 3 are waiting for VMs
2. Hour later, first 2 builds finish, second 2 builds start and last one is still waiting
3. Another hour later second 2 builds finish (just in time before timeout) and last, 5th, build fails because of timeout

So we end up with 4 finished builds and one failed because of timeout.

Anyway, I'm happy to change the wording as I think we both agree that the issue is there.